File: examples/audio.cpp
#include <math.h>
#include <stdio.h>

#include <emlite/emlite.hpp>

using namespace emlite;

int main() {
    emlite::init();
    auto AudioContext = Val::global("AudioContext");
    if (!AudioContext.as<bool>()) {
        printf("No global AudioContext, trying "
               "webkitAudioContext\n");
        AudioContext = Val::global("webkitAudioContext");
    }

    printf("Got an AudioContext\n");
    auto context    = AudioContext.new_();
    auto oscillator = context.call("createOscillator");

    printf("Configuring oscillator\n");
    oscillator.set("type", "triangle");
    oscillator.get("frequency").set("Value", 261.63);


    auto doc  = Val::global("document");
    auto body = doc.call("getElementsByTagName", "body")[0];
    auto btn  = doc.call("createElement", "BUTTON");
    btn.set("textContent", "Play!");
    emlite_val_inc_ref(oscillator.as_handle());
    btn.call(
        "addEventListener",
        "click",
        Val::make_fn([=](auto params) -> Val {
            printf("Playing\n");
            oscillator.call(
                "connect", context.get("destination")
            );
            oscillator.call("start", 0);
            printf("All done!\n");
            return Val::undefined();
        })
    );
    body.call("appendChild", btn);
}

File: rust/emlite/examples/audio.rs
use emlite::*;

fn main() {
    #[allow(non_snake_case)]
    let mut AudioContext = Val::global("AudioContext");
    if !AudioContext.as_bool() {
        println!("No global AudioContext, trying webkitAudioContext");
        AudioContext = Val::global("webkitAudioContext");
    }

    println!("Got an AudioContext");
    let context = AudioContext.new(&[]);
    let oscillator = context.call("createOscillator", &[]);

    println!("Configuring oscillator");
    oscillator.set("type", "triangle");
    oscillator.get("frequency").set("value", Val::from(261.63)); // Middle C

    Val::global_this().set("oscillator", oscillator);
    Val::global_this().set("context", context);

    let document = Val::global("document");
    let elem = document.call("createElement", &argv!["BUTTON"]);
    elem.set("textContent", "Click");
    let body = document.call("getElementsByTagName", &argv!["body"]).at(0);
    elem.call(
        "addEventListener",
        &argv![
            "click",
            Val::make_fn(|_| {
                let oscillator = Val::global("oscillator");
                let context = Val::global("context");
                println!("Playing");
                oscillator.call("connect", &argv![context.get("destination")]);
                oscillator.call("start", &argv![0]);
                println!("All done!");
                Val::undefined().as_handle()
            })
        ],
    );
    body.call("appendChild", &argv![elem]);
}
